crates/noirc_frontend/src/monomorphization/mod.rs

```diff
@@ -686,12 +741,39 @@ impl<'interner> Monomorphizer<'interner> {
             }
         }
 
-        self.try_evaluate_call(&func, &return_type).unwrap_or(ast::Expression::Call(ast::Call {
-            func,
-            arguments,
-            return_type,
-            location,
-        }))
+        let mut block_expressions = vec![];
+
+        let is_closure = self.is_function_closure(call.func);
+        if is_closure {
+            let extracted_func: ast::Expression;
+            let hir_call_func = self.interner.expression(&call.func);
+            if let HirExpression::Lambda(l) = hir_call_func {
+                let (setup, closure_variable) = self.lambda_with_setup(l, call.func);
+                block_expressions.push(setup);
+                extracted_func = closure_variable;
+            } else {
+                extracted_func = *original_func;
+            }
+            func = Box::new(ast::Expression::ExtractTupleField(
+                Box::new(extracted_func.clone()),
+                1usize,
+            ));
+            let env_argument = ast::Expression::ExtractTupleField(Box::new(extracted_func), 0usize);
+            arguments.insert(0, env_argument);
+        } else {
+            func = original_func.clone();
+        };
```

Comment:
Can we break this out into another function?

Reply:
I am not fully sure if this logic wouldn't be more seriously improved anyway in the next reform, and currently it updates several local variables(`block_expressions`, `arguments`), so I'd prefer to leave the refactoring out for the next PR: otherwise I agree